File: python/lesson 5/trie.py
# https://leetcode.com/problems/implement-trie-prefix-tree/

class TrieNode:
    def __init__(self):
        self.children = [None] * 26
        self.isWord = False

    def set(self, char, node):
        self.children[ord(char) - ord('a')] = node

    def get(self, char):
        return self.children[ord(char) - ord('a')]

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node  = self.root
        for c in word:
            if not node.get(c):
                node.set(c, TrieNode())
            node = node.get(c)
        node.isWord = True
        
    def search(self, word: str) -> bool:
        node = self.findNode(word)
        return node is not None and node.isWord

    def startsWith(self, prefix: str) -> bool:
        node = self.findNode(prefix)
        return node is not None
    
    def findNode(self, prefix: str) -> TrieNode:
        node  = self.root
        for c in prefix:
            node = node.get(c)
            if not node: return None
        return node        

File: ruby/lesson 5/Trie.rb
# https://leetcode.com/problems/implement-trie-prefix-tree/

class Trie
    def initialize
        @root = TrieNode.new
    end

    def insert(word)
        node = @root
        word.each_char do |c|
            node[c] ||= TrieNode.new
            node = node[c]
        end
        node.is_word = true
    end

    def search(word)
        node = find_node(word)
        !node.nil? && node.is_word
    end

    def starts_with(prefix)
        node = find_node(prefix)
        !node.nil?
    end

    private

    def find_node(prefix)
        node = @root
        prefix.each_char do |c|
            return nil if node[c].nil?
            node = node[c]
        end
        node
    end

    class TrieNode
        attr_accessor :is_word

        def initialize()
            @children = Array.new(26) # a to z
            @is_word = false
        end

        def [](c)
            @children[c.ord - 'a'.ord]
        end

        def []=(c, value)
            @children[c.ord - 'a'.ord] = value
        end
    end
end
